bson/codec_options.py

```diff
@@ -33,29 +33,58 @@ def _raw_document_class(document_class):
     return marker == _RAW_BSON_DOCUMENT_MARKER
 
 
+class TypeCodecBase(object):
+    @property
+    def python_type(self):
+        """The Python type to be converted into something serializable."""
+        raise NotImplementedError
+
+    @property
+    def bson_type(self):
+        """The BSON type to be converted into our own type."""
+        raise NotImplementedError
+
+    def transform_bson(self, value):
+        """Convert the given BSON value into our own type."""
+        raise NotImplementedError
+
+    def transform_python(self, value):
+        """Convert the given Python object into something serializable."""
+        raise NotImplementedError
+
+
 class TypeRegistry(object):
-    def __init__(self, encoder_map=None, decoder_map=None):
-        self._encoder_map = (encoder_map or {}).copy()
-        self._decoder_map = (decoder_map or {}).copy()
+    def __init__(self, *args):
+        self.__args = args
+        self._encoder_map = {}
+        self._decoder_map = {}
+        for arg in args:
+            if not isinstance(arg, TypeCodecBase):
+                raise TypeError(
+                    "Expected instance of %s" % (TypeCodecBase.__name__,))
+            try:
+                python_type = arg.python_type
+            except NotImplementedError:
+                pass
+            else:
+                self._encoder_map[python_type] = arg.transform_python
```

Comment:
Should we do something special if  `arg.transform_python` (or `arg.transform_bson`) raises NotImplementedError?

Reply:
I was planning to include some validation logic into `TypeCodecBase.__init__` to ensure that if a codec advertises support for a type on the encode or decode side, it also implements the corresponding conversion. However, I have no reliable way of figuring out an argument I can use to call the transformer function to test if it raises that exception. The best I could do was call `testvalue = python_type()` followed by a try/except on `arg.transform_python(testvalue)` but that assumes that the default constructor of `python_type` returns a properly initialized object which might not be true for every custom type (though it is true for all types I tried - decimal, int64, set, frozenset). 
What do you think would be the best way to test something like this?

Reply:
Thanks for explaining that! Now I understand why we can't do this validation. The user will get an exception the first time it's actually used anyway so this SGTM.